### ISSUE TITLE
[Q] River networks for real time application

### ISSUE THREAD

**Issue by trsh** [dandrino/terrain-erosion-3-ways#7]
Your thoughts on optimizing river_networks for real time application, a game. Where chunks are generated as player walks on the map.  Possible/not?

**Comment by dandrino** [comment]
In the way it was written, no. You could plausibly create something in which each chunk has a believable erosion pattern going on, but each chunk will not be logically coherent with respect to each other in regards to the rules of erosion (e.g. you may get patterns in which a river does not drain into a hydraulic sink).

The closest I think you can get with respect to hydraulic erosion (assuming you intent to create something that looks realistic) is an algorithm that generates tiles on the fly, but has some global context or state that is managing how each tile is generated. This is sort of intermediate between completely independent tile generation (like Perlin noise), and a completely global generation process (what I've done in this repo). As an example using real life, if the algorithm generated the mouth of the Amazon river, further tiles would have to be generated with the knowledge that an entire continent's worth of water would need to feed into it. I have no idea what this coordinating process would look like, and getting it to achieve results that are consistent, believable, and fast would be challenging.

A higher level question to consider is what you are gaining and sacrificing by using infinite tile generation. You gain benefits in the form of an infinitely explorable world, but the costs are that you have a more limited toolbox to work with (which could mean terrain that is less interesting). I don't know what constraints motivate your decision making, but it's useful to consider which option provides more of a net benefit than the other.

**Comment by trsh** [comment]
I was so obsessed with performance I did not think of tileabillity lol. Thank you for the insight!

**Comment by trsh** [comment]
@dandrino any advice or peace of knowledge how to get h. erosion in unlimited procedural height maps?

**Comment by dandrino** [comment]
I don't know of anything yet that can accomplish what you are looking for. As mentioned above, it would likely have to be an on-the-fly system that is dependent on global state management rather than a purely decentralized approach. I don't know what this algorithmic strategy would look like in practice, and my suspicions are that it would likely be a complicated and non-obvious solution.

The theoretical challenge is that hydraulic erosion introduces dependencies between one tile and another that make it difficult to generate using an unlimited and on-the-fly approach. The volume of water in one river segment will depend on the size and structure of the drainage basin feeding into it, and therefore you cannot generate those two independently of one another. 

Now, it may be the case that through some mathematical sleight of hand you can map values from an infinitely generated domain (e.g. Perlin noise) onto those that resemble hydraulic erosion, but I have no idea what that would look like (perhaps ML would work in this case, but that's a difficult tool to use). Given that hydraulic erosion is driven by a complex set of differential equations, I suspect that a neat, closed-form formula will likely be out of reach.

tl;dr; Currently I don't know of any way to generate the effects demoed in this repo for unlimited procedural height maps.

**Comment by trsh** [comment]
Tnx
